### DISCUSSION TITLE
Verify bidder authenticity on chain - using seller as bidder PKH oracle

### DISCUSSION THREAD

**Discussion by nazrhom** [mlabs-haskell/hydra-auction#301] [Ideas]
## Introduction

This item stems from our internal slack discussion in https://mlabs-corp.slack.com/archives/C04C3ASUFEF/p1683191319843989?thread_ts=1683057601.835759&cid=C04C3ASUFEF

To summarise, the issue we found is that: 

_We can only use the verifySignature primitive to verify that a certain VKey has signed the piece of data (or, the SKey corresponding to the VKey has signed the data, to be more precise). But we have no way to link this to the PKH on-chain._

## Idea

To overcome this issue, we need to establish a way to link a certain VK with a certain PKH on chain.

We can do this by having the seller generate a signature for each bidder witnessing the match. We can include this new signature in the standing bid datum, so we can prove the link between VK and PKH on chain.

## Implementation details

For each valid deposit, the seller will:
- derive bidderPKH and bidderVK from deposit
- sign a tuple `(auctionId, bidderPKH, bidderVK)`
- make it available to the bidder

--- 

Extend `BidTerms` with additional fields:
```haskell
 bidderVK :: BuiltinByteString
 bidderSignature :: BuiltinByteString
 sellerSignature :: BuiltinByteString
```

bidderSignature will be a tuple `(auctionId, bidPrice)`
sellerSignature is as described above

---

Extend `AuctionTerms` with additional field:

```haskell
sellerVK :: BuiltinByteString
````

---

newBid redeemer can now check that:
- bidderSignature matches `(hydraHeadId terms, bidPrice newBidTerms)` and is signed by `bidderVK newBidTerms`
- sellerSignature matches `(hydraHeadId terms, bidderPKH newBidTerms, bidderVK newBidTerms)` and is signed by `sellerVK terms`

**Comment by GeorgeFlerovsky** [comment]
Looks good, but I would reinforce it by including the bidder's PKH in the tuple that the bidder signs with his VK when submitting a new bid:

`(auction ID, bidder PKH, bid amount)`

Just in case to prevent the seller from lying about PKH<->VK without the bidder's consent. When the bidder signs this tuple, he is saying "I agree to bid this amount in this auction, such that the auction lot will go to this PKH if I win."

The PKH in the bidder's tuple and the seller's tuple must match.

**Reply by nazrhom** [reply]
That is a good point, I'll add that and update the spec with this new approach.

The only part that needs some more thinking would be: how exactly does the seller make the signed tuple `(auctionId, bidderPKH, bidderVK)` available to the bidder?

One solution could be to store this on the delegate servers, and give them a new endpoint so the bidder can query if their deposit is "accepted" and they can start bidding, but maybe we don't need to implement the whole solution for the demo?

**Reply by uhbif19** [reply]
> The only part that needs some more thinking would be: how exactly does the seller make the signed tuple (auctionId, bidderPKH, bidderVK) available to the bidder?

This is covered by Platform server. See #300
